src/ProjectOrigin.WalletSystem.Server/Services/WalletService.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Threading.Tasks;
+using Google.Protobuf;
+using Grpc.Core;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using ProjectOrigin.WalletSystem.Server.Database;
+using ProjectOrigin.WalletSystem.Server.HDWallet;
+using ProjectOrigin.WalletSystem.Server.Models;
+
+namespace ProjectOrigin.WalletSystem.Server.Services;
+
+[Authorize]
+public class WalletService : ProjectOrigin.WalletSystem.V1.WalletService.WalletServiceBase
+{
+    private readonly string _endpointAddress;
+    private readonly ILogger<WalletService> _logger;
+    private readonly UnitOfWork _unitOfWork;
+    private readonly IHDAlgorithm _hdAlgorithm;
+
+    public WalletService(ILogger<WalletService> logger, UnitOfWork unitOfWork, IHDAlgorithm hdAlgorithm, IOptions<ServiceOptions> options)
+    {
+        _endpointAddress = options.Value.EndpointAddress;
+        _logger = logger;
+        _unitOfWork = unitOfWork;
+        _hdAlgorithm = hdAlgorithm;
+    }
+
+    public override async Task<V1.WalletSectionReference> CreateWalletSection(V1.CreateWalletSectionRequest request, ServerCallContext context)
+    {
+        var subject = context.GetSubject();
+
+        var wallet = await _unitOfWork.WalletRepository.GetWallet(subject);
+
+        if (wallet is null)
+        {
+            var key = _hdAlgorithm.GenerateNewPrivateKey();
+            wallet = new Wallet(Guid.NewGuid(), subject, key);
+            await _unitOfWork.WalletRepository.Create(wallet);
+        }
+
+        int nextPosition = await _unitOfWork.WalletRepository.GetNextWalletPosition(wallet.Id);
+
+        var section = new WalletSection(Guid.NewGuid(), wallet.Id, nextPosition, wallet.PrivateKey.Derive(nextPosition).PublicKey);
```

Comment:
This will always go together (get the next position and then create the WalletSection). Could this be delegated to the repository? Or if `Wallet` has its Sections loaded, then a method could be added to `Wallet` to create the next section

Reply:
Yes might be a good addition to move this to the repository, then it might also be easier to solve the concurrency issue.